util/topgen/templates/plic_all_irqs_test.c.tpl

```diff
@@ -0,0 +1,246 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+${gencmd}
+<%
+irq_peripheral_names = sorted({p.name for p in helper.irq_peripherals})
+enabled_peripherals = ['uart']
+%>\
+
+#include "sw/device/lib/base/freestanding/limits.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+% for n in irq_peripheral_names:
+<%
+  comment = '' if n in enabled_peripherals else '// '
+%>\
+${comment}#include "sw/device/lib/dif/dif_${n}.h"
+% endfor
+#include "sw/device/lib/handler.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_framework/test_main.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+% for p in helper.irq_peripherals:
+<%
+  comment = '' if p.name in enabled_peripherals else '// '
+%>\
+${comment}static dif_${p.name}_t ${p.inst_name};
+% endfor
+static dif_rv_plic_t plic;
+static const top_earlgrey_plic_target_t kHart = kTopEarlgreyPlicTargetIbex0;
+
+/**
+ * Flag indicating which peripheral is under test.
+ *
+ * Declared volatile because it is referenced in the main program flow as well
+ * as the ISR.
+ */
+static volatile top_earlgrey_plic_peripheral_t peripheral_expected;
+
+/**
+ * Flags indicating the IRQ expected to have triggered and serviced within the
+ * peripheral.
+ *
+ * Declared volatile because it is referenced in the main program flow as well
+ * as the ISR.
+ */
+% for n in irq_peripheral_names:
+<%
+  comment = '' if n in enabled_peripherals else '// '
+%>\
+${comment}static volatile dif_${n}_irq_t ${n}_irq_expected;
+${comment}static volatile dif_${n}_irq_t ${n}_irq_serviced;
+% endfor
+
+/**
+ * Provides external IRQ handling for this test.
+ *
+ * This function overrides the default external IRQ handler in
+ * `sw/device/lib/handler.h`.
+ */
+void handler_irq_external(void) {
+  // Find which interrupt fired at PLIC by claiming it.
+  dif_rv_plic_irq_id_t plic_irq_id;
+  CHECK(dif_rv_plic_irq_claim(&plic, kHart, &plic_irq_id) == kDifRvPlicOk,
+        "dif_rv_plic_irq_claim failed");
+
+  // Check if it is the right peripheral.
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == peripheral_expected,
+        "Interurpt from incorrect peripheral: exp = %d, obs = %d",
+        peripheral_expected, peripheral);
+<%text>
+#define PERIPHERAL_ISR(_peripheral, _handle, _plic_start_irq)                 \
```

Comment:
So, the question to ask here: why is this jammed into a generated file, instead of being in a "real" header? It doesn't seem to depend on anything in the template, so maybe it makes sense to inline it into where it's used below, or put it in its own header.

Whatever can be hoisted out of this header probably should.

Also, is there a reason for the leading underscore names? We don't have specfic style guidance here, so we should be consistent with the rest of the project.

Reply:
After thinking about this more from our discussion earlier, I think the style guides are confusing here. Specifically, defining these macros right before they are used, and undefining them immediately after is directly inline with the [Google C++ style guide](https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros), for which we mostly follow. However, there are a couple sentences in our [(OpenTitan) style guide](​​https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) that seem to toss this guidance out the window, and enable, as @mcy suggests, hoisting the macros to a public header: "_In contrast to the recommendation in the Google C++ style guide, exporting macros as part of the public API is allowed in C code. A typical use case is a header with register definitions._", though the exact use case here is different.

Regardless of my style guide naivete, since this is autogenerated code, I think it makes sense to keep this inlined in the Mako template for loop context you make use of below (as @mcy suggests), and just do away with the macro entirely. This would naturally also remove the leading underscore names :) Likewise with the other macro blocks.

Reply:
> I did not see any real benefit to moving these to a separate header since they are not going to be used in any other tests, hence no need to make these public.

There is a benefit: they are now real C files with real indexing. The templates are very hard to read, and once we move to Bazel we will probably stop checking generated code in in a lot of cases. Please factor the macros out into a header.

> The leading underscore helps distinguish macro args from the identifiers in the rest of the test code.

I know, but the rest of the project doesn't do this. If you feel strongly, send me a PR to the style guide and fixes to make the entire project consistent with this style. I can see the benefit, but OT is small enough that consistency is important.

Reply:
Ok. Moved the macros to headers, and removed leading underscores.